api/src/main/java/org/jenkinsci/plugins/workflow/graph/BlockEndNode.java

```diff
@@ -34,19 +37,29 @@
  * @see BlockStartNode
  */
 public abstract class BlockEndNode<START extends BlockStartNode> extends FlowNode {
-    private final START start;
+    private transient START start;
+    private final String startId;
 
     public BlockEndNode(FlowExecution exec, String id, START start, FlowNode... parents) {
         super(exec, id, parents);
         this.start = start;
+        startId = start.getId();
     }
 
     public BlockEndNode(FlowExecution exec, String id, START start, List<FlowNode> parents) {
         super(exec, id, parents);
         this.start = start;
+        startId = start.getId();
     }
 
     public START getStartNode() {
+        if (start == null) {
+            try {
+                start = (START) getExecution().getNode(startId);
+            } catch (IOException x) {
+                Logger.getLogger(BlockEndNode.class.getName()).log(Level.WARNING, null, x);
```

Comment:
If <code>getExecution().getNode(startId)</code> fails, we may be spammed by warnings here if a code above is recurrent. Maybe makes sense to add a flag, which blocks logs after first failure(?)